feat(messenger): introduce rudimentary discord import tool

0x-r4bbit opened the pull request:
**NOTE: I've put up 2 commits here, the second one is just for `make generate` and will be squashed after review is done. Only the first one needs to be reviewed.**

This commit aims to implement the first parts of the discord import
tool. It's not complete as there are more things that need to be handle
by it, but as of this commit, it introduces the basic functionality
that:

1. Let's clients request a discord import
2. Creates a community + its categories/channels derived from import
   data
3. Converts discord messages to users messages and saves them to the DB
4. Downloads discord message author avatar assets and stores them as
   base64 encoded payloads
5. Initializes the imported community by setting up filters and
   notifications
6. During all of the above, it sends progress updates to clients so they
   can show users meaningful UIs

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (6)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 1ace6aec | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2826/1/) | 2022-08-26 15:47:19 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220826-154310-1ace6a-pr2826.zip) |
| :heavy_check_mark: | 1ace6aec | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2826/1/) | 2022-08-26 15:47:49 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220826-154315-1ace6a-pr2826.zip) |
| :heavy_check_mark: | 1ace6aec | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2826/1/) | 2022-08-26 15:49:41 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220826-154310-1ace6a-pr2826.aar) |
| | | | | | | |
| :heavy_check_mark: | 9ebfa853 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2826/2/) | 2022-08-26 16:09:50 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220826-160511-9ebfa8-pr2826.zip) |
| :heavy_check_mark: | 9ebfa853 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2826/2/) | 2022-08-26 16:10:40 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220826-160508-9ebfa8-pr2826.aar) |
| :heavy_check_mark: | 9ebfa853 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2826/2/) | 2022-08-26 16:11:01 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220826-160508-9ebfa8-pr2826.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b8a578c8 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2826/3/) | 2022-08-26 19:46:05 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220826-194353-b8a578-pr2826.zip) |
| :heavy_check_mark: | b8a578c8 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2826/3/) | 2022-08-26 19:46:37 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220826-194353-b8a578-pr2826.zip) |
| :heavy_check_mark: | b8a578c8 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2826/3/) | 2022-08-26 19:47:29 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220826-194349-b8a578-pr2826.aar) |
| | | | | | | |
| :heavy_check_mark: | 49f04144 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2826/4/) | 2022-08-26 21:43:34 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220826-214118-49f041-pr2826.zip) |
| :heavy_check_mark: | 49f04144 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2826/4/) | 2022-08-26 21:43:36 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220826-214115-49f041-pr2826.zip) |
| :heavy_check_mark: | 49f04144 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2826/4/) | 2022-08-26 21:44:45 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220826-214115-49f041-pr2826.aar) |

Macias131987 commented on appdatabase/migrationsprevnodecfg/bindata.go:
```diff
@@ -137,7 +137,7 @@ func _0001_appDownSql() (*asset, error) {
 
-	info := bindataFileInfo{name: "0001_app.down.sql", size: 356, mode: os.FileMode(0644), modTime: time.Unix(1644328579, 0)}
+	info := bindataFileInfo{name: "0001_app.down.sql", size: 356, mode: os.FileMode(0664), modTime: time.Unix(1660816022, 0)}
 	a := &asset{bytes: bytes, info: info, digest: [32]uint8{0xb5, 0x25, 0xa0, 0xf8, 0x7d, 0x2d, 0xd, 0xcf, 0x18, 0xe4, 0x73, 0xc3, 0x95, 0xf5, 0x24, 0x20, 0xa9, 0xe6, 0x9e, 0x1d, 0x93, 0xe5, 0xc5, 0xad, 0x93, 0x8f, 0x5e, 0x40, 0xb5, 0x30, 0xaa, 0x25}}
```
Fernando macias

Macias131987 reviewed (approved):
Fernando macias Saldana

0x-r4bbit committed:
feat(messenger): introduce rudimentary discord import tool

This commit aims to implement the first parts of the discord import
tool. It's not complete as there are more things that need to be handle
by it, but as of this commit, it introduces the basic functionality
that:

1. Let's clients request a discord import
2. Creates a community + its categories/channels derived from import
   data
3. Converts discord messages to users messages and saves them to the DB
4. Downloads discord message author avatar assets and stores them as
   base64 encoded payloads
5. Initializes the imported community by setting up filters and
   notifications
6. During all of the above, it sends progress updates to clients so they
   can show users meaningful UIs

0x-r4bbit committed:
generate bindata

0x-r4bbit commented on protocol/discord/assets.go:
```diff
@@ -0,0 +1,32 @@
+	}
+	imageBase64 := fmt.Sprintf("data:image/%s;base64,%s", mime, encBuf)
+	return payload, &imageType, imageBase64, nil
+}
```
This function will be used to download discord message author avatar assets.
Right now it returns there actual image payload + the type + the base64 string, but as of now I'm only storing the base64 string (don't think we need both, payload + base64).

However I've noticed we store payload + base64 for message images as well. Shall I do the same for discord author images?

0x-r4bbit commented on protocol/discord/types.go:
```diff
@@ -1,17 +1,24 @@
 )
 
-type ErrorType uint
+type ErrorCodeType uint
```
Renamed this so we can have `WarningType` and `ErrorType` as enum below.
Reason I want those for the enum below is so we can have `Error()` and Warning` as function types.

0x-r4bbit commented on protocol/discord/types.go:
```diff
@@ -57,10 +92,102 @@ type ImportError struct {
+		Message: message,
+		Code:    WarningType,
+	}
+}
```
Here are `Error()` and `Warning()`

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1832,29 +1833,20 @@ func (m *Messenger) ExtractDiscordDataFromImportFiles(filesToImport []string) (*
-				Code:    2,
-				Message: err.Error(),
-			}
+			errors[fileToImport] = discord.Error(err.Error())
```
Here's how `Error()` and Warning()` are used as convenience APIs

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+			return
+		}
+
+		importProgress.UpdateTaskProgress(discord.CommunityCreationTask, 0.5)
```
Setting the progress to `0.5` is somewhat arbitrary here. I split up the `CommunityCreationTask` into a few parts and extracting the data accounts for `50%` of the work.

Happy to change this to something more reasonable.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+		}
+
+		importProgress.CommunityID = discordCommunity.IDString()
+		importProgress.UpdateTaskProgress(discord.CommunityCreationTask, 0.75)
```
Same thing here. `0.75` is arbitrary. Happy to change this. (but also: in reality this happens so fast, that these updates are barely noticable in the UI)

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+				processedCategoriesIds[category.ID] = addedCategory.CategoryId
+			}
+			m.publishImportProgress(importProgress)
+		}
```
Notice how the categories are created sequentially. Theoretically, we can do this in parallel, but judging by what I've experienced using real data, this is rather fast. Anyways, if anyone thinks it's worth it to pursue I'm happy to do that in a follow up PR. Right now I don't see the need.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+		var discordMessagesToSave []*protobuf.DiscordMessage
+		var chatsToSave []*Chat
+		authorAvatarUrls := map[string]string{}
+		// messageAssetsToDownload := map[string][]*protobuf.DiscordMessageAttachment{}
```
Whoops, left over. Need to remove this.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+			}
+			importProgress.UpdateTaskProgress(discord.ImportMessagesTask, float32(len(processedMessages))/float32(totalMessageCount)*0.90)
+			m.publishImportProgress(importProgress)
+		}
```
Processing the channels however, I think we can definitely do in parallel. Something I'd like to explore as optimization once the core functionality has landed.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+				discordMessagesToSave = append(discordMessagesToSave, discordMessage)
+				processedMessages = append(processedMessages, messageToSave)
+			}
+			importProgress.UpdateTaskProgress(discord.ImportMessagesTask, float32(len(processedMessages))/float32(totalMessageCount)*0.90)
```
Oh here, I'm multiplying by `0.9` so the progress bar goes to 90%. Reason being, we still need to save the processed messages which will take time. So we don't want it to be 100% yet.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+			m.publishImportProgress(importProgress)
+			return
+		}
+		importProgress.UpdateTaskProgress(discord.ImportMessagesTask, 0.95)
```
Same here. Arbitrary value, as we don't get updates in between save calls.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1921,3 +1914,389 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+			return
+		}
+
+		importProgress.UpdateTaskProgress(discord.InitCommunityTask, 0.5)
```
Also arbitrary, and the following ones for `InitCommunityTask` as well

0x-r4bbit commented:
Closing this as it's stale
